### ISSUE TITLE
Some issues afaik 

### ISSUE THREAD

**Issue by guthasaibharathchandra** [guthasaibharathchandra/PS-KD-Pytorch#1]
1) https://github.com/guthasaibharathchandra/PS-KD-Pytorch/blob/b7347b390b04fd342b391c8a6b3f992c118ece47/main.py#L454
   preds_teacher here is only the logits , they should be softmax normalized before passing to criterion_student in line 457 (and now the linear combination inside StudentLoss should work)

2) https://github.com/guthasaibharathchandra/PS-KD-Pytorch/blob/b7347b390b04fd342b391c8a6b3f992c118ece47/main.py#L460
I'm not sure how the flags in this 'if block' are used.  Ideally we should backpropagate through our student head, teacher head, learnable parameters always, and depending on the flag, we should backpropagate through the backbone via student or teacher

3) https://github.com/guthasaibharathchandra/PS-KD Pytorch/blob/b7347b390b04fd342b391c8a6b3f992c118ece47/models/preact_resnet.py#L245
This line will only make sure the learnable parameters are normalized when they are initialized. But when we train, these change and so we need to make sure these are normalized in the forward() function.  Also I feel there's no need for line 268 i.e no need to normalize teacher_out. Instead, we need to add F.normalize(teacher_pred) after line 269 (i.e after  teacher_pred = self.learnable_params(teacher_out))

Please let me know If any of the above is an issue and if I misunderstood something.

**Comment by Supersamu** [comment]
Hi, this all should be fixed with my most recent commit. 

Re 1: Thank you, I overlooked that, now the softmax is taken before linear combination with the ground truth

Re 2: We now have three optimizers, one for the heads, and one for backpropagating through student backbone and one for backpropagating through teacher backbone. I know they are equivalent, but we might potentially have different learning rates for both operations.

Re 3: This I think you have missed. Line 516 in main.py normalizes the learnable parameters

**Comment by Supersamu** [comment]
About the last comment again: I am not sure what the purpose of the normalization of the teacher predictions are if we then use a softmax later. The slides say this: _teacher prediction = normalized similarity vector of normalized feature vector and normalized learnable targets_ ,so maybe we do need to L2-normalize three times and then do the softmax?

**Comment by guthasaibharathchandra** [comment]
Yeah, that's what I was referring to. May be we can confirm this with yassine tomorrow.
